### DISCUSSION TITLE
Predicting the Gradient of DeepONet Solutions

### DISCUSSION THREAD

**Discussion by LanPeng-94** [lululxvi/deepxde#1589] [Q&A]
Hi,

I use the `y_pred = model.predict(data.test_x)`, and the shape of y_pred is 200*841. Now, I want to get the gradient of each DeepONet solution. If the inputs of the trunk are x1 and x2, I write the following code,

```
def dydx(inputs, outputs):
    x_loc = inputs[1]
    du_x = tf.gradients(outputs, x_loc)[0][:, 0:1]
    print("du_x_Grad shape {}".format(du_x.shape))
    return du_x
```

`dy_x = model.predict(data.test_x, operator=dydx)`
However, the shape of dy_x is 841*1. It doesn't seem to have obtained the gradient corresponding to each predicted value, what should I do.

Best regards,

Peng

**Comment by lululxvi** [comment]
Do you use DeepONet or DeepONetCartesianProd?

**Reply by LanPeng-94** [reply]
Hi,

I use DeepONetCartesianProd to build my surrogate model. 
I guess it seems to only get the gradient of the last function. Because the shape of y_pred is 200*841, when I use `tf.gradients(outputs, x_loc)[0][:, 0:1]`, it can not obtain the gradient for each discrete functions. Based on this, I modify my code,

```
y_pred = model.predict(data.test_x)

dy_xs = []
for i in range(200):
    def dydx(inputs, outputs):
        x_loc = inputs[1]
        du_x = tf.gradients(outputs[i:i+1, :], x_loc)[0][:, 0:1]
        # print("du_x_Grad shape {}".format(du_x.shape))
        return du_x
    dy_x = model.predict(data.test_x, operator=dydx).T
    print(dy_x.shape)
    dy_xs.append(dy_x)
    
dy_xs = np.array(dy_xs).astype(np.float32)
print("dy_xs shape {}".format(dy_xs.shape))
```

I want to know if my understanding is correct. Is there any problem with my code?

Best regards,

Peng

**Reply by lululxvi** [reply]
You can read the code here to know how to compute the gradients

https://github.com/lululxvi/deepxde/blob/2236965af761173c120a27d1615c9b2847b3053c/deepxde/data/pde_operator.py#L234
